src/main/java/com/gtnewhorizons/modularui/api/forge/FluidTankLong.java

```diff
@@ -0,0 +1,230 @@
+package com.gtnewhorizons.modularui.api.forge;
+
+import static com.google.common.primitives.Ints.saturatedCast;
+
+import java.io.IOException;
+
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidRegistry;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidTankInfo;
+import net.minecraftforge.fluids.IFluidTank;
+
+/**
+ * A fluid tank with long capacity and stored amount
+ * 
+ * @author BlueWeabo
+ */
+public class FluidTankLong implements IFluidTank {
+
+    private Fluid fluid;
+    private long storedAmount;
+    private long capacity;
+    private FluidStack internal;
+    private boolean locked;
+    private int lastFluidAmountInStack;
+    private NBTTagCompound tag;
+
+    public FluidTankLong(Fluid fluid, long capacity, long amount) {
+        this.fluid = fluid;
+        this.storedAmount = amount;
+        this.capacity = capacity;
+        if (fluid == null) return;
+        internal = new FluidStack(fluid, saturatedCast(amount));
+    }
+
+    public FluidTankLong(Fluid fluid, long capacity) {
+        this(fluid, capacity, 0);
+    }
+
+    public FluidTankLong(long capacity) {
+        this((Fluid) null, capacity);
+    }
+
+    public FluidTankLong(IFluidTank tank) {
+        this(tank.getFluid() != null ? tank.getFluid().getFluid() : null, tank.getCapacity(), tank.getFluidAmount());
```

Comment:
This is not the right way to handle composition tank; `FluidTankLong` only knows its originally stored fluid and does not propagate manipulations. So original tank does not know what fluid manipulation was done. You can check it with MetaTE machines not saving fluids. I suspect this is also the cause of the bug you sent on Discord.

So what we should do instead is
1. Create interface `IFluidTankLong` that extends `IFluidTank`
2. Make `FluidTankLong` implement `IFluidTankLong` and make other classes to use it
3. Create another class `FluidTankLongDelegate` or smth that takes `IFluidTank` as constructor and delegates all the calls for `IFluidTankLong` to it

Reply:
The thing is that I don't use a IFluidTank when I got that bug, so it seems kind of unlikely for that?

Reply:
I see, I'll look at it later

Reply:
I prepared for it mostly so it should be fine.

Reply:
MetaTE machines not saving fluids is still a thing and needs addressing

Reply:
One example is
1. MetaTE machine creates `FluidSlotWidget` from `FluidStackTank`. The tank contains molten iron.
2. In the constructor of `FluidSlotWidget`, tank is wrapped by `FluidTankLong` and the wrapper is passed to widget.
3. `FluidTankLong` constructor remembers currently stored fluid, and immediately forgets the passed tank.
4. Player drains fluid from `FluidSlotWidget`. It's propagated to `FluidTankLong` and modification is made there, but not to the original `FluidStackTank`. `FluidTankLong` contains null fluid, but `FluidStackTank` still thinks it contains molten iron.
5. Player fills fluid to `FluidSlotWidget`. It's propagated to `FluidTankLong` and modification is made there, but not to the original `FluidStackTank`. `FluidTankLong` contains molten gold, but `FluidStackTank` still thinks it contains molten iron.
6. When saving machine data to NBT, the machine thinks the original `FluidStackTank` knows the actual stored fluid and creates saved data from it. But actually it doesn't know what has happened to it at all, so molten iron is saved.

Reply:
I won't be able to fix this today as I will soon be going out to do my driving practical test, for which i have no clue how much time will take. So if possible look into if its still an issues now that FluidTankLongDelegate exists.